### ISSUE TITLE
Add NetBSD version of CLI

### ISSUE THREAD

**Issue by krytarowski** [dotnet/coreclr#3178]
@ellismg FYI

```
Restoring BuildTools...
Unsupported OS NetBSD detected. Downloading ubuntu-x64 tools
cat: /etc/*-release: No such file or directory
cat: /etc/*-release: No such file or directory
cat: /etc/*-release: No such file or directory
cat: /etc/*-release: No such file or directory
.NET CLI will be downloaded from https://dotnetcli.blob.core.windows.net/dotnet/dev/Binaries/1.0.0.000973/dotnet-ubuntu-x64.1.0.0.000973.tar.gz
```

What are the next steps to add NetBSD there?

**Comment by ellismg** [comment]
We have to have a working CoreFX and CoreCLR before we can add things.  Is all the native code in CoreCLR and CoreFX building cleanly yet?  Do the CoreCLR pal tests pass?

**Comment by jasonwilliams200OK** [comment]
I don't think this makes any sense at this point. CoreCLR master does not build on NetBSD without patches at this point, 20% PAL tests are still failing when build CoreCLR with patches, the entire managed side of CoreFX (aka .NET Core BCL) is yet to be ported and CoreFX System.Native still have one final patch (in addition to couple of already opened PRs).

**Comment by krytarowski** [comment]
93 PAL tests to go

**Comment by ellismg** [comment]
FYI, when you have native CoreFX and CoreCLR building, the best way to work around the bootstrapping problems with the CLI is to set `BUILDTOOLS_OVERRIDE_RUNTIME` to a folder that contains all the native code you've built (for both CoreFX and CoreCLR).  `init-tools.sh` will use this folder to stomp over all the native CLR bits after restoring a toolset, which will let you replace the linux binaries with the NetBSD versions.  At that point, I expect you should be able to build mscorlib.dll on NetBSD.

**Comment by krytarowski** [comment]
Thank you for tips -- I will try to make use of them.

**Comment by krytarowski** [comment]
Practically all PAL tests are passing on NetBSD (modulo pending patches). Tomorrow I will switch to CoreFX.

**Comment by jasonwilliams200OK** [comment]
@ellismg, apparently setting `BUILDTOOLS_OVERRIDE_RUNTIME` is not enough given the current state of build structure when we have `dotnet` binary mandatory requirement, which is not available for FreeBSD/NetBSD.

I tried compiling CLI, but the scripts rely on `dotnet-$osname-x64.latest.tar.gz` which is not available for FreeBSD/NetBSD. Using `osx` as an override doesn't work either because the the downloaded tar contains a native binary, which is not compatible.

Are there plans to open source full set of tools available on GitHub so the new platforms don't get blocked? Can we expect something soon'ish? :)

**Comment by krytarowski** [comment]
CoreFX is still awaiting for the PCL elimination, until then I'm porting LLILC to NetBSD. It should be a good initial verification of CoreCLR.

**Comment by ellismg** [comment]
> Are there plans to open source full set of tools available on GitHub so the new platforms don't get blocked? Can we expect something soon'ish? :)

To a first approximation, everything has been open sourced.  There aren't any tools that I'm aware of that we are holding back that would help in bringing a platform online.

What hasn't been done is for someone to sit down and understand the exact process that needs to be followed to bring a new platform online.  We've developed the technology we've needed organically as we've run into issues (e.g. `BUILDTOOLS_OVERRIDE_RUNTIME` which was helpful for cases where you could build the native bits of coreclr and corefx but we didn't have NuGet packages for your distro yet) but no one has tried to actually bring up a new platform like FreeBSD or NetBSD in a while.

I (and others on the team) do want to make this something that's both possible and easy.  However, there's some higher priority stuff we need to land first to really get the CLI on solid footing before I can go look at this stuff.

So, I hear what you are saying and I agree it's an issue.  I'm sorry what we have today isn't sufficient and I'm also sorry that we have more pressing issues we need to address for the current set of platforms CoreCLR already runs on.  But rest assured that this is something I want to make better and I'm trying to fit it into my schedule.

**Comment by jasonwilliams200OK** [comment]
Thanks @ellismg for the detailed response, we appreciate your efforts to solidify the tools based on the existing set platform. We should certainly acknowledge what you guys have achieved and delived over the period of an year, and that is absolutely brilliant effort! :+1: 

I didn't knew about this dependency situation. The last time when i worked on FreeBSD, `dotnet/cli` project was non-existent and CoreFX was depending on using `aspnet/dnx`. When I saw build is switching to `dotnet/cli` I didn't knew that would be a show stopper. Today I booted up FreeBSD to find CoreFX tests status (in response to https://github.com/dotnet/buildtools/pull/477) and discovered this blocker. This could also mean that in past couple of weeks, nobody has tried building CoreFX managed assemblies on FreeBSD.

**Comment by krytarowski** [comment]
@ellismg Thank you for your reply.. well there is still something to do.. LLILC has been ported (via pkgsrc-wip) to NetBSD... I think I will try with Pyjiton too :smile:

**Comment by stephentoub** [comment]
FWIW, this shouldn't be a show-stopper.  For a long time we did the managed builds on Windows and then copied the built binaries over to the other system... that still works!  So you should absolutely be able to make forward progress that way if you're so inclined.

**Comment by jasonwilliams200OK** [comment]
@stephentoub, that is a good idea, I will give it a shot.

**Comment by nbyouri** [comment]
I am eagerly awaiting for NetBSD support of coreclr!

**Comment by jasonwilliams200OK** [comment]
@nbyouri, us as well; please let us know if you are willing to participate too. :)

**Comment by nbyouri** [comment]
Sure, I'm willing to help!

**Comment by krytarowski** [comment]
CoreCLR 100% PAL tests pass
CoreFX managed code tests: Total: 90143 Errors: 0 Failed: 2 Skipped: 0 Time: 9998 (50k crashing the runtime)

Are we close to the point to get CLI bootstrap?
